#107

packpaul opened the pull request.

packpaul committed:
#107 Implement WITH NO AUTOUPDATE for CREATE GRAIN in CelestaSQL
Implementation with a parser test

packpaul committed:
#107 Implement WITH NO AUTOUPDATE for CREATE GRAIN in CelestaSQL
Test for 'schema with no autoupdate' added

inponomarev reviewed (changes requested):
Your branch auto-building failed on ratcheting phase (there are more checkstyle/findbugs warnings in your branch than in latest master). Please make Jenkins green before we can proceed with the code review

packpaul committed:
#107 Implement WITH NO AUTOUPDATE for CREATE GRAIN in CelestaSQL
Adjustments after styling check

inponomarev commented on celesta-sql/src/main/javacc/CelestaParser.jj:
```diff
@@ -236,6 +241,7 @@ Grain parseGrainPart(GrainPart grainPart) :
 
    Token t;
    Token doc = null;
+   boolean val;
```
init val to true in declaration, no need to init it on line 261

inponomarev commented on celesta-sql/src/main/javacc/CelestaParser.jj:
```diff
@@ -118,6 +118,11 @@ PARSER_END(CelestaParser)
     <K_GRAIN_VERSION: "VERSION">
 }
 
+<SCHEMA_OPTIONS_STATE> TOKEN:
```
I would prefer not to create new lexer state here!

Better just rename `TABLE_OPTIONS_STATE` to `OPTIONS_STATE` and reuse it both for table and grain options. It all will be shorter

inponomarev reviewed (changes requested):
generally ok but please simplify the JavaCC specification

packpaul committed:
#107 Implement WITH NO AUTOUPDATE for CREATE GRAIN in CelestaSQL
Adjustments after code review

packpaul commented:
@inponomarev Done.

inponomarev commented on celesta-sql/src/main/java/ru/curs/celesta/dbutils/DbUpdater.java:
```diff
@@ -114,6 +114,9 @@ public void updateDb() {
             // Выполняем итерацию по гранулам.
             boolean success = true;
             for (Grain g : grains) {
+                if (! g.isAutoupdate()) {
```
Why did you introduce two separate flags, `isAutoupdate` and `lock` on `Grain`?

You set `isAutoupdate`, but you test `lock` in `SchemaAutoupdateTest`.

I think you should do the following: 

1. Remove `lock` flag
2. `decideToUpgrade` should use `isAutoupdate` flag, not `lock`
3. Remove `if (! g.isAutoupdate()) continue` on line 117: we got all the upgrading/skipping logic in `decideToUpgrade` method

packpaul replied:
1. Removing `if (! g.isAutoupdate()) continue` on line 117 would mean we'd get grains **with no autoupdate** be inserted into DB since branch
```
if (gi == null) {
    insertGrainRec(g);
    success = updateGrain(g, connectionPool) & success;
}
```
would take place. Is that what we need?

2. lock flag was introduced on purpose in order to add test case 2. in [#107](https://github.com/CourseOrchestra/celesta/issues/107) but I think I need to set lock status over ISchemaCursor, so I'll remove the lock flag.

inponomarev commented on celesta-sql/src/test/java/ru/curs/celesta/score/ParserTest.java:
```diff
@@ -63,6 +63,27 @@ private void parseAndSaveCsqlScript(String csqlScript, File grainPath, String gr
     }
   }
+  
+  @Test
```
This test is insufficient. You check that you parse the `WITH NO AUTOUPDATE` without any errors, but you never `assertFalse(grain.isAutoupdate)`. 

It is also desirable to check for correct parsing the `WITH AUTOUPDATE` clause (`assertTrue(grain.isAutoupdate)`) and the default behaviour (no `WITH` clause, assertTrue(grain.isAutoupdate)).

See examples of such tests in `test1`...`test4` (yeah I know, this is awful, but they are very old tests ;-) )

Together with `SchemaAutoupdateTest` this is going to be OK.

packpaul replied:
A separate test `SchemaAutoupdateTest.testScoreParsing()` added.

packpaul committed:
#107 Implement WITH NO AUTOUPDATE for CREATE GRAIN in CelestaSQL
Grain.lock property removed. SchemaAutoupdateTest rewritten to use
locking over ISchemaCursor

packpaul committed:
Merge remote-tracking branch 'origin/dev' into #107

packpaul committed:
#107 Implement WITH NO AUTOUPDATE for CREATE GRAIN in CelestaSQL
SchemaAutoupdateTest.testScoreParsing() added
